arch/xtensa/src/esp32/esp32_spiflash.c

```diff
@@ -1470,6 +1551,219 @@ static int IRAM_ATTR esp32_readdata_encrypted(
+          else
+            {
+              memcpy(wbuf, buffer, n);
+            }
+        }
+      else
+        {
+          n = SPI_FLASH_ENCRYPT_UNIT_SIZE;
+          wbuf = (uint8_t *)buffer;
+          addr = offset;
+        }
+
+      ret = esp32_writedata_encrypted(priv, addr, wbuf,
+                                      SPI_FLASH_ENCRYPT_UNIT_SIZE);
+      if (ret < 0)
+        {
+          ferr("esp32_writedata_encrypted failed ret=%d\n", ret);
+          break;
+        }
+
+      offset += n;
+      nbytes -= n;
+      buffer += n;
+      wbytes += n;
+    }
+
+  return wbytes;
+}
+
+/****************************************************************************
+ * Name: esp32_spiflash_work
+ *
+ * Description:
+ *   Do SPI Flash operation, cache result and send semaphore to wake up
+ *   blocked task.
+ *
+ * Input Parameters:
+ *   p - SPI Flash work arguments
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP32_SPI_FLASH_SUPPORT_PSRAM_STACK
+static void esp32_spiflash_work(void *p)
+{
+  struct spiflash_work_arg *work_arg = (struct spiflash_work_arg *)p;
+
+  if (work_arg->op_code == SPIFLASH_OP_CODE_WRITE)
+    {
+      work_arg->ret = esp32_writedata(work_arg->op_arg.priv,
+                                      work_arg->op_arg.addr,
+                                      work_arg->op_arg.buffer,
+                                      work_arg->op_arg.size);
+    }
+  else if (work_arg->op_code == SPIFLASH_OP_CODE_READ)
+    {
+      esp32_set_read_opt(work_arg->op_arg.priv);
+      work_arg->ret = esp32_readdata(work_arg->op_arg.priv,
+                                     work_arg->op_arg.addr,
+                                     work_arg->op_arg.buffer,
+                                     work_arg->op_arg.size);
+    }
+  else if (work_arg->op_code == SPIFLASH_OP_CODE_ERASE)
+    {
+      work_arg->ret = esp32_erasesector(work_arg->op_arg.priv,
+                                        work_arg->op_arg.addr,
+                                        work_arg->op_arg.size);
+    }
+  else if (work_arg->op_code == SPIFLASH_OP_CODE_ENCRYPT_READ)
+    {
+      esp32_set_read_opt(work_arg->op_arg.priv);
+      work_arg->ret = esp32_readdata_encrypted(work_arg->op_arg.priv,
+                                               work_arg->op_arg.addr,
+                                               work_arg->op_arg.buffer,
+                                               work_arg->op_arg.size);
+    }
+  else if (work_arg->op_code == SPIFLASH_OP_CODE_ENCRYPT_WRITE)
+    {
+      work_arg->ret = esp32_writeblk_encrypted(work_arg->op_arg.priv,
+                                               work_arg->op_arg.addr,
+                                               work_arg->op_arg.buffer,
+                                               work_arg->op_arg.size);
+    }
+  else
+    {
+      ferr("ERROR: op_code=%d is not supported\n", work_arg->op_code);
+    }
+
+  sem_post(&work_arg->sem);
```

Comment:
```suggestion
  nxsem_post(&work_arg->sem);
```
`sem_post` should not be called within the Kernel since it sets `errno`.